api/common/crypto.cc

```diff
@@ -0,0 +1,266 @@
+    nop = 0;
+    if (!EVP_CipherFinal_ex(&aes, NULL, &nop) || nop != 0)
+        goto exit;
+
+    // Write result to user's string
+    encrypted->reserve(result_size);
+    for (int i = 0; i < result_size; i++)
+        *encrypted += result[i];
+    ret = true;
+exit:
+    EVP_CIPHER_CTX_cleanup(&aes);
+    return ret;
+}
+
+bool DecAes256CbcSha512(const std::string &key,
+                        const std::string &encrypted,
+                        std::string *clear) {
+    if (key.length() != AES256_KEY_SIZE ||
+        encrypted.length() < AES_IV_SIZE + SHA512_SIZE + AES_BLOCK_SIZE ||
+        encrypted.length() % AES_BLOCK_SIZE != 0)
+        return false;
+    bool ret = false;
+    const uint8_t *iv =
+        reinterpret_cast<const uint8_t *>(encrypted.c_str());
+    const uint8_t *k =
+        reinterpret_cast<const uint8_t *>(key.c_str());
+    const uint8_t *enc_blocks = iv + AES_IV_SIZE;
+    int blocks_len = encrypted.length() - AES_IV_SIZE;
+    uint32_t us;
+    int s;
+    EVP_CIPHER_CTX aes;
+    EVP_MD_CTX *sha;
+
+    std::unique_ptr<uint8_t> clear_result(new uint8_t[blocks_len]);
+    uint8_t *clear_sha = clear_result.get();
+    uint8_t *clear_payload = clear_sha + SHA512_SIZE;
+    int clear_payload_len;
+    uint8_t computed_sha[SHA512_SIZE];
+    uint8_t pad_len;
+
+    // Prepare AES cipher with included IV
+    EVP_CIPHER_CTX_init(&aes);
+    EVP_CipherInit_ex(&aes, EVP_aes_256_cbc(), NULL, k, iv, 1);
+    EVP_CIPHER_CTX_set_padding(&aes, 0);
+
+    // Decrypt data
+    if (!EVP_CipherUpdate(&aes, clear_result.get(), &s, enc_blocks, blocks_len)
+        || s != blocks_len)
+        goto exit;
+
+    s = 0;
+    if (!EVP_CipherFinal_ex(&aes, NULL, &s) || s)
+        goto exit;
+
+    pad_len = clear_result.get()[blocks_len - 1];
+
+    if (pad_len > AES_BLOCK_SIZE)
+        goto exit;
+
+    clear_payload_len = blocks_len - SHA512_SIZE - pad_len;
+
+    // Compute SHA of decrypted data
+    sha = EVP_MD_CTX_create();
+    EVP_DigestInit_ex(sha, EVP_sha512(), NULL);
+    EVP_DigestUpdate(sha, clear_payload, clear_payload_len);
+    EVP_DigestFinal_ex(sha, computed_sha, &us);
+    if (us != SHA512_SIZE)
+        goto exit;
+    EVP_MD_CTX_destroy(sha);
+
+    // Check that SHA is correct
+    if (memcmp(clear_sha, computed_sha, SHA512_SIZE))
+        goto exit;
+
+    // Copy decrypted data
+    clear->reserve(clear_payload_len);
+    for (int i = 0; i < clear_payload_len; i++)
+        *clear += clear_payload[i];
+    ret = true;
+exit:
+    EVP_CIPHER_CTX_cleanup(&aes);
+    return ret;
+}
+
+#undef AES_IV_SIZE
+#undef AES_BLOCK_SIZE
+#undef AES256_KEY_SIZE
+#undef SHA512_SIZE
+}  // namespace Crypto
+
+/* Simple unit test.
+ * Build with:
+ * g++ -std=c++11 -lcrypto -DUNIT_TESTS crypto.cc -o crypto
+ */
+#ifdef UNIT_TESTS
+static int CheckAes256CbcSha512(const std::string msg) {
+    std::string key, enc, dec;
+    for (uint8_t i = 0; i < 32; i++) key += i;
```

Comment:
so the key is {0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32} ?